### PR TITLE
Visualisation: Allow specifying Agent shapes in agent_portrayal

### PR DESCRIPTION
**Pull request by rmhopkins4**
This PR allows specifying an `"shape"` in the `agent_portrayal` dictionary used by matplotlib component of the Solara visualisation. In short, it allows you represent an Agent in any [matplotlib marker](https://matplotlib.org/stable/api/markers_api.html), by adding a "shape" key-value pair to the `agent_portrayal` dictionary.

This is especially useful when you're using the default shape drawer for grid or continuous space.

For example:
```Python
def agent_portrayal(cell):
    return {
        "color": "blue"
        "size": 5,
        "shape": "h"  # marker is a hexagon!
    }
```
Partial resolution of #2164.

### Examples

In Conway's Game of Life life, you can now use black squares for the agents, as how it's traditionally displayed:

```Python
def agent_portrayal(cell):
    return {
        "color": "white" if cell.alive else "black",
        "size": 45,
        "shape": "s" # marker is a square
    }
```
image

You can also conditionally modify the shape, just like you can with the color and size:

```Python
COP_COLOR = "#000000"
AGENT_QUIET_COLOR = "#648FFF"
AGENT_REBEL_COLOR = "#FE6100"
JAIL_COLOR = "#808080"
JAIL_SHAPE = "s"


def citizen_cop_portrayal(agent):
    portrayal = {
        "Shape": "circle",
        "x": agent.pos[0],
        "y": agent.pos[1],
        "Filled": "true",
    }

    if isinstance(agent, Citizen):
        color = (
            AGENT_QUIET_COLOR if agent.condition == "Quiescent" else AGENT_REBEL_COLOR
        )
        if agent.jail_sentence:
            color = JAIL_COLOR
            shape = JAIL_SHAPE
        else:
            shape = "."
        color = JAIL_COLOR if agent.jail_sentence else color
        portrayal["color"] = color
        portrayal["shape"] = shape
        portrayal["size"] = 15.0

    else:
        assert isinstance(agent, Cop)
        portrayal["color"] = COP_COLOR
        portrayal["shape"] = "."
        portrayal["size"] = 25.0

    return portrayal
```
image

### CODE REVIEW

File: mesa/visualization/components/matplotlib.py
Comment (rht) [inline mesa/visualization/components/matplotlib.py:53]: I'm not sure of the performance implication of plotting each points separately. What about https://stackoverflow.com/a/43528356?

File: mesa/visualization/components/matplotlib.py
Comment (rht) [inline mesa/visualization/components/matplotlib.py:28]: It seems that Altair natively supports this: https://github.com/vega/altair/issues/1181.

File: mesa/visualization/components/matplotlib.py
Comment (rmhopkins4) [reply mesa/visualization/components/matplotlib.py:53]: Points are grouped by marker (shape) before plotting, like in the link provided.

File: mesa/visualization/components/matplotlib.py
Comment (rmhopkins4) [reply mesa/visualization/components/matplotlib.py:28]: Altair space_drawer does not support continuous space. To have continuous space w/ multiple shapes altair must either support continuous space or matplotlib must support multiple shapes.

File: mesa/visualization/components/matplotlib.py
Comment (rmhopkins4) [reply mesa/visualization/components/matplotlib.py:53]: one way to potentially speed the grouping up is to use numpy.
if the x, y, s, c, markers arrays were np arrays, masks could be used on each np.unique() marker

Comment (rmhopkins4) [reply mesa/visualization/components/matplotlib.py:53]: But I understand not wanting to introduce dependencies if not strictly necessary

**Comment by rht** [comment]
Implementation LGTM, thank you for implementing this feature. I could test it on the Epstein civil violence model if I have the time. But if you don't mind, showing that the jailed citizens as square in that example would be helpful. You can used the code from https://github.com/projectmesa/mesa-examples/pull/158 and undo the experimental data collector part.

File: mesa/visualization/components/matplotlib.py
Comment (rht) [reply mesa/visualization/components/matplotlib.py:28]: I think we should do both, with the Altair one in a separate PR. In your code, it falls back to the default code branch when shape is not present, so it's fine.

**Comment by rmhopkins4** [comment]
image
Here is an example of jailed citizens as squares.
While working on this I made matplotlib.py provide default visualization values so that users only need to provide overrides

**Comment by rmhopkins4** [comment]
> image Here is an example of jailed citizens as squares. While working on this I made matplotlib.py provide default visualization values so that users only need to provide overrides

Matplotlib's 's' and 'c' (size, color) accept no values (all default), one value (all the same), and lists exactly as long as the number of points provided. 
When the list lengths were not aligned problems arose.

File: mesa/visualization/components/matplotlib.py
Comment (rht) [inline mesa/visualization/components/matplotlib.py:39]: Can't use assertion here. Needs to raise an exception because this would be disabled by the Python `-O` flag (to disable assertions).

**Comment by EwoutH** [comment]
Thanks a lot for this PR! I will review it somewhere tomorrow.

File: mesa/visualization/components/matplotlib.py
Comment (rht) [inline mesa/visualization/components/matplotlib.py:41]: The error needs to show the public name of these attributes: x, y, size, color, marker instead of their Matplotlib argument names.

File: mesa/visualization/components/matplotlib.py
Comment (rht) [inline mesa/visualization/components/matplotlib.py:85]: Matplotlib's `tab:blue` used in their default plot looks nicer than the default `b` (`blue`).

Comment (rht) [reply mesa/visualization/components/matplotlib.py:85]: See https://matplotlib.org/stable/gallery/color/named_colors.html#tableau-palette.

File: mesa/visualization/components/matplotlib.py
Comment (rht) [inline mesa/visualization/components/matplotlib.py:87]: According to https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.scatter.html, the default is actually `"o"`.

**Comment by EwoutH** [comment]
@rht could you give a short summary what you already have sufficiently reviewed and what could use a second pair of eyes?

**Comment by rht** [comment]
Actually, this PR is already LGTM, and I didn't merge because you still wanted to review. If any, maybe the perf consideration of an array of defaults instead of a single default value scalar, but this is minor because any loop done without calling Matplotlib many times is going to be fast.

**Review by rht — APPROVED** [review]

**Comment by EwoutH** [comment]
Thanks. I would have expected this feature taking less complexity / lines of code, let me take a closer look tomorrow morning.

**Comment by EwoutH** [comment]
Okay, I'm largely good with this PR. It's a bit weird https://github.com/matplotlib/matplotlib/issues/11155 was never implemented, but since the only thing changes for our users is that they can now _optionally_ pass a matplotlib shape in a `portrayal` dict (right @rht?), we can always implement the back-end later in a more (performance) efficient way.

I would like the documentation around this PR to improve a little:
- Update the docstring to note that a `"shape"` can be passed, and add a link to the list with allowed values (I think it's [this one](https://matplotlib.org/stable/api/markers_api.html))
- Update the first comment of this PR with the screenshot, one or two more examples, and a link to 
- Mention that you can pass "shape" in the [visualization_tutorial](https://mesa.readthedocs.io/en/latest/tutorials/visualization_tutorial.html) ([source](https://github.com/projectmesa/mesa/blob/main/docs/tutorials/visualization_tutorial.ipynb)).

**Comment by rht** [comment]
> Okay, I'm largely good with this PR. It's a bit weird https://github.com/matplotlib/matplotlib/issues/11155 was never implemented, but since the only thing changes for our users is that they can now optionally pass a matplotlib shape in a portrayal dict (right @rht?), we can always implement the back-end later in a more (performance) efficient way.

Yes, it's optional and is not a breaking change.

While it's good to update the docstring now, I think the tutorial change can happen in a separate PR. Otherwise it makes Mesa too bureaucratic to contribute to (that every new feature needs to be in the tutorial in the same PR).

**Comment by EwoutH** [comment]
Yeah I see your perspective and also considered that. But otherwise I'm afraid it doesn't get done. Especially with our user base changes need to be (very) visible.

@rmhopkins4 let me know if you want to do these things yourself, or otherwise allow us to do it within this PR.

**Comment by rmhopkins4** [comment]
> Yeah I see your perspective and also considered that. But otherwise I'm afraid it doesn't get done. Especially with our user base changes need to be (very) visible.
> 
> @rmhopkins4 let me know if you want to do these things yourself, or otherwise allow us to do it within this PR.

You should have the ability to edit the PR already, no?

**Comment by EwoutH** [comment]
Yes I do, but I don't prefer to edit other peoples PRs without their explicit permission :)

**Comment by EwoutH** [comment]
@rmhopkins4 Could you let me know if you want to take up (any of) the remaining points yourself, or if you want us as maintainers to do it?

**Comment by rmhopkins4** [comment]
I can take up the remaining points. If I understand correctly it's just documentation-related.

**Comment by rmhopkins4** [comment]
Is there a docstring that currently mentions the available options for the `agent_portrayal`? They seem to just state that `agent_portrayal` lets you specify agents' display options but not how.

**Comment by EwoutH** [comment]
I think you’re right, it’s weird we don’t have that documented.

It might have gone lost with the transition from the old to the new visualization.

I think this would be the most logical place to add it:
https://github.com/projectmesa/mesa/blob/3cf1b76ca3a157c8a064b60130755a86e73718be/mesa/visualization/solara_viz.py#L106
But is also shows how important PR, release and tutorial documentation is - that’s how users discover new features!

**Comment by EwoutH** [comment]
@rmhopkins4 are you done working on it? It looks good!

File: mesa/visualization/components/matplotlib.py
Comment (rht) [reply mesa/visualization/components/matplotlib.py:55]: I want to add that NumPy has a lot of overhead. It is faster only for large arrays.

**Comment by EwoutH** [comment]
> Here is an example of jailed citizens as squares.

@rmhopkins4 I'm ready to merge. Could you add the portrayal code you used to the first PR post?

**Comment by EwoutH** [comment]
Congratulations of getting your first PR into Mesa @rmhopkins4!

**Comment by EwoutH** [comment]
@rmhopkins4 would you like to update a few example models to use the shapes?

If so, this issue includes a list of models already using the new visualization:
- https://github.com/projectmesa/mesa-examples/issues/154

**Pull request merged**
